### PR TITLE
Platform specific file copying

### PR DESCRIPTION
**Pull request by juhhov**
## Status
**READY**

## Description
Revert back to platform specific file copying as the current implementation is not robust.

In non windows platforms file is now created with flags:
`os.O_CREAT | os.O_TRUNC | os.O_RDWR | os.O_SYNC`
with the exception of arm machines which stay:
`os.O_CREAT | os.O_TRUNC | os.O_RDWR | os.O_DIRECT`  

http://man7.org/linux/man-pages/man2/open.2.html

https://jira.arm.com/browse/IOTSYSTOOL-1290

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def read(fname):
           "mbed-ls>=1.5.1,==1.*",
           "six",
           "pyserial",
-          "pyOCD"
+          "pyOCD==0.12.0"
```
[the remark this answers is not in the document]
Comment (jupe) [reply setup.py]: even better would be separate pr

File: mbed_flasher/flashers/FlasherMbed.py [hunk 2]
```diff
@@ -201,14 +201,22 @@ def copy_file(self, source, destination):
...
-                new_file.write(aux_source)
-                new_file.flush()
+            new_file = open(destination, "wb", 0)
+            new_file.write(aux_source)
+            new_file.flush()
+            try:
+                # Intermittently this step fails in Raspberry PI.
+                # It doesn't seem to affect the result of copy, ignore the fail.
                 os.fsync(new_file.fileno())
+            except (IOError, OSError):
```
Comment (jupe) [inline mbed_flasher/flashers/FlasherMbed.py]: is this safe thing to do..?

Comment (juhhov) [reply mbed_flasher/flashers/FlasherMbed.py]: With current knowledge it is.

Comment (jupe) [reply mbed_flasher/flashers/FlasherMbed.py]: ok

**Comment by jupe** [comment]
why py3 fails?

File: mbed_flasher/flashers/FlasherMbed.py [hunk 3]
```diff
@@ -199,17 +202,34 @@ def copy_file(self, source, destination):
...
+                self._copy_file_windows(source, destination)
+            else:
+                self._copy_file(aux_source, destination)
+        except (IOError, OSError, subprocess.CalledProcessError):
             self.logger.exception("File couldn't be copied")
             raise FlashError(message="File couldn't be copied",
                              return_code=EXIT_CODE_OS_ERROR)
 
+    def _copy_file_windows(self, source, destination):
+        command = "copy \"{}\" \"{}\"".format(os.path.abspath(source), destination)
```
Comment (jupe) [inline mbed_flasher/flashers/FlasherMbed.py]: why this `os.path.abspath()` ?

File: mbed_flasher/flashers/FlasherMbed.py [hunk 4]
```diff
@@ -199,17 +202,34 @@ def copy_file(self, source, destination):
...
+        self.logger.debug("Copying with command: {}".format(command))
+        subprocess.check_call(command)
```
Comment (jupe) [inline mbed_flasher/flashers/FlasherMbed.py]: this is still an problem, need some change

Comment (juhhov) [reply mbed_flasher/flashers/FlasherMbed.py]: Why is it a problem?

Comment (YifeiZuo01) [reply mbed_flasher/flashers/FlasherMbed.py]: I don't understand the purpose to start a new command interpreter, would mind explain a bit?
```
C:\>cmd /?
Starts a new instance of the Windows XP command interpreter
```

Comment (juhhov) [reply mbed_flasher/flashers/FlasherMbed.py]: It is one way to avoid using shell=True, which we do not want to do for security reasons.

**Comment by RomanSaveljev** [comment]
> the current implementation is not robust

Once this change is in, what is the plan to measure whether the system is more robust.. less robust.. same robust?

**Comment by juhhov** [comment]
> the current implementation is not robust
That comment is targeted to #143, which does not pass acceptance. The changes this PR brings compared to v0.9.1 are described in description.
Currently there is no other way to measure robustness than acceptance tests and opentmi results to some extent.

**Pull request merged**